### PR TITLE
Do not release react events to the pool.

### LINKED ISSUE #2: Can event handlers yield and get the event object invalidated too early?
React re-uses event objects in a pool, so yielding a thread that processes an event object could potentially break it.  I'm not sure this is actually an issue, but I would like to be more confident that it's not.

See also: https://bitbucket.org/wuzzeb/react-flux/src/2be756ffa5d60bdc15f767f4006f0e145efcb39e/src/React/Flux/PropertiesAndEvents.hs?at=default&fileviewer=file-view-default#PropertiesAndEvents.hs-372

### LINKED ISSUE #3: Eliminate uses of deepseq.
`deepseq` used to be called in one instance to get the old implementation of `preventDefault`, `stopPropagation` to work (see https://bitbucket.org/wuzzeb/react-flux/src/2be756ffa5d60bdc15f767f4006f0e145efcb39e/src/React/Flux/PropertiesAndEvents.hs?at=default&fileviewer=file-view-default#PropertiesAndEvents.hs-372 for details); this is no longer a valid reason, even though there is another issue with that (see #2)

I think there are uses where I have not found the reason yet.

I would like to have all occurrances of `deepseq` either explained or removed.

### LINKED ISSUE #26: better types for event handlers
A lot of the code has a type variable called `handler` or `eventHandler`, which can only be one of these two values:

```haskell
type ViewEventHandler = [SomeStoreAction]
type StatefulViewEventHandler state = state -> ([SomeStoreAction], Maybe state)
```

(both may be changed to `IO` soon, but don't let that confuse you, that's not the point here.)

The code would become much more accessible if the type variables would be kind-restricted so that they can only be one of these two types.  I feel this should be easy?

### LINKED ISSUE #34: Can we do better than UnoverlappingEq?
`UnoverlappingEq` works, but it's not very nice.  Perhaps we can change the code so `Eq` can be used instead?

### PR DESCRIPTION
**Pull request by fisx**
Fixes #2.  Fixes #3.  Fixes #34.  Fixes #26.  Related: #15, #8.

- [x] Better implementation for `stopPropagation`, `preventDefault`.
- [x] Make `eventHandler` type kinded (more type safety).
- [x] Call `<evt>.persist()` implicitly on all react events to allow for threaded event handlers.
- [x] Remove `NFData` constraints from action, props, state types.
- [x] Remove some dead code from `Outdated`.

(Haddocks will be updated in a separate PR.)

(un-squashed backup of this branch: 2-concurrency-and-events-2)

### CODE REVIEW

**Comment by fisx** [comment]
> persistReactEvent raw h = js_persistReactEvent raw >> isThreadContinueAsync _ >> h raw

@luite like this?

**Comment by luite** [comment]
I'm not sure that's the right thing to check.

But it looks like there's no way to make a synchronous thread async (I
wanted to follow the unmask/mask thing, but haven't added that yet). Maybe
this is a good excuse to just do that.

**Comment by fisx** [comment]
- [x] the renamed new types should contain a haddock comment that references löh paper about agda terminology.
- [x] rename ViewEventHandler to EventHandler, Statful...
- [x] remove all NFData constraints?  none of them should be necessary any more after the changes in this PR.

File: react-hs/src/React/Flux/Internal.hs [hunk 1]
```diff
@@ -100,13 +104,17 @@ instance Show HandlerArg where
 -- | Marker class to make things less polymorphic (we likely only ever need 'ViewEventHandler',
 -- 'StatefulViewEventHandler', so just throwing around unconstrained type variables is a little
 -- confusing).
-class IsEventHandler handler
+data EHandler s = EHView | EHState s
```
Comment (fisx) [inline react-hs/src/React/Flux/Internal.hs]: do we need the `s` parameter here?  isn't it always `*`, so we can keep it implicit?

**Comment by fisx** [comment]
(unsquashed branch in https://github.com/liqula/react-hs/pull/25/commits/1b6fb1ac37b9db6cd20cc0407de03de91ebd9787)

**Comment by fisx** [comment]
@basvandijk @aische as I talked have talked to you about this library in other contexts, I'd be interested in your opinion if you have any.  Thanks :)

**Comment by fisx** [comment]
(backup of unsquashed branch: https://github.com/liqula/react-hs/pull/25/commits/bb22fd571f6121a44717401ad37f16220e63189d)

**Comment by rjmk** [comment]
I think one option to make `EventHandlerType` closed would be to move it, `PropertyOrHandler` (but not `PropertyOrHandler_` into `PropertiesAndEvents`. This would require a few more imports in various places (importing `Store` into `PropertiesAndEvents` and `PropertiesAndEvents` into `DOM`). I'm not sure if it's an overall win

Is the idea behind using the type family rather than a GADT to save the pain of the data constructors (and preferring additional annotations / applications)? I'm on board with that, just wanted to check that I followed the reasoning correctly there.

On TypeApplications, maybe enable them in the cabal file at this point? They see a lot of use and it's quite strange seeing an extension being used a lot but not being in the cabal file, and then some others occuring in the cabal file (e.g. DataKinds).

I approve of the changes to `preventDefault` etc. I'll have a think about how that might be tweaked in the longer term to allow more event modifications.

**Comment by fisx** [comment]
Thanks for the feedback, and sorry for the long response time.

> I think one option to make EventHandlerType closed would be to move it, PropertyOrHandler (but not PropertyOrHandler_ into PropertiesAndEvents. This would require a few more imports in various places (importing Store into PropertiesAndEvents and PropertiesAndEvents into DOM). I'm not sure if it's an overall win

Agreed with the last point.  :-)  I'll add this idea to the haddocks, though.

> Is the idea behind using the type family rather than a GADT to save the pain of the data constructors (and preferring additional annotations / applications)?

I think yes, the idea was to avoid constructor noise.  (@divipp?)

> On TypeApplications, maybe enable them in the cabal file at this point? They see a lot of use and it's quite strange seeing an extension being used a lot but not being in the cabal file, and then some others occuring in the cabal file (e.g. DataKinds).

done: 10ec697bccef2ff541ecbd94bfc1a2ad5b5328c3.

> I approve of the changes to preventDefault etc. I'll have a think about how that might be tweaked in the longer term to allow more event modifications.

Yes, it would be great to have a better solution here.  I've been pointed out some improvements to this though that I will add to the documentation and examples today.

**Comment by NicoleRauch** [comment]
What I do not understand at all is what simpleHandler is actually doing here and why you'd want that... A function signature for simpleHandler might be a good start (why is that not part of the PR?), a function definition might be even more helpful (same question applies), and in case this is some old stuff that you're just reusing here, a named import might help me find the relevant code...

Just to give you an idea of how basic the issues are that I'm struggling with :-)

**Comment by fisx** [comment]
@NicoleRauch I don't understand your questions.

```
$ git grep -A1 -Hn 'simpleHandler ::'
react-hs/src/React/Flux/PropertiesAndEvents.hs:284:simpleHandler :: h -> (h, [EventModification])
react-hs/src/React/Flux/PropertiesAndEvents.hs-285-simpleHandler = (, [])
```

There is also a haddocks string when you open that module, namely the one I just updated for you in https://github.com/liqula/react-hs/pull/25/commits/6604ae1577eda6c2315987018d8dab276cfd2bbe.

Then `git blame react-hs/src/React/Flux/PropertiesAndEvents.hs | grep simpleHandler` shows you that the function has been added in 7b365efac5da6218c20e421ba825d5fd0af852e7.

Does this help?  (If so, what was the problem, and which part of the answer *did* help?)

thanks-

**Pull request merged**